docs/src/writing_good_rules.md

```diff
@@ -103,16 +103,197 @@ In fact, any number of scalar arguments is supported, as is returning a tuple of
+    y = addone!_cr(a)
+    function addone!_cr_pullback(ȳ)
+        return (NO_FIELDS, ones(length(a)))
+    end
+    return y, addone!_cr_pullback
+end
+```
+the gradient can be evaluated:
+```julia
+julia> gradient(addone!_cr, a)
+([1.0, 1.0, 1.0],)
+```
+
+!!! note "Why is there an extra `addone!_cr` function defined?"
+    This is only done for illustration purposes so that it is possible to follow the steps in the REPL.
+    When `gradient` is called in `Zygote` for a function with no `rrule` defined, a backward pass for the function call is generated and cached.
+    When `gradient` is called for the second time on the same function signature, the backward pass is reused without checking whether an an `rrule` has been defined between the two calls to `gradient`.
+    If an `rrule` is defined before the first call to `gradient` it should register the rule and use it, but that prevents comparing what happens before and after the `rrule` is defined.
+
+#### Exception handling
+
+Zygote does not support differentiating through `try`/`catch` statements.
+For example, differentiating through
+```julia
+function exception(x)
+    try
+        return x^2
+    catch e
+        println("could not square input")
+        throw(e)
+    end
+end
+exception_cr(x) = exception(x)
+```
+does not work
+```julia
+julia> gradient(exception, 3.0)
+ERROR: Compiling Tuple{typeof(exception),Int64}: try/catch is not supported.
+```
+without an `rrule` defined
+```julia
+function ChainRulesCore.rrule(::typeof(exception_cr), x)
+    y = exception_cr(x)
+    function exception_cr_pullback(ȳ)
+        return (NO_FIELDS, 2*x)
+    end
+    return y, exception_cr_pullback
+end
+```
+
+```julia
+julia> gradient(exception_cr, 3.0)
+(6.0,)
+```
+
+
+#### Loops
+
+Julia runs loops fast.
+Unfortunately Zygote differentiates through loops slowly.
+So, for example, computing the mean squared error by using a loop
+```julia
+function mse(y, ŷ)
+    N = length(y)
+    s = 0.0
+    for i in 1:N
+        s +=  (y[i] - ŷ[i])^2.0
+    end
+    return s/N
+end
+mse_cr(y, ŷ) = mse(y, ŷ)
+```
+takes a lot longer to AD through
+```julia
+julia> y = rand(30)
+julia> ŷ = rand(30)
+julia> @btime gradient(mse, y, ŷ)
+  39.561 μs (1007 allocations: 65.33 KiB)
+```
+than if we supply an `rrule`,
+```julia
+function ChainRules.rrule(::typeof(mse_cr), x, x̂)
+    output = mse_cr(x, x̂)
+    function mse_cr_pullback(ȳ)
+        N = length(x)
+        g = (2 ./ N) .* (x .- x̂) .* ȳ
+        return (NO_FIELDS, g, -g)
+    end
+    return output, mse_cr_pullback
+end
+```
+which is much faster
+```julia
+julia> @btime gradient(mse_cr, y, ŷ)
+  1.293 μs (18 allocations: 1.09 KiB)
+```
+
+#### Inplace accumulation
+
+Inplace accumulation of gradients is slow in `Zygote`. The issue, demonstrated in the
+folowing example, is that the gradient of `getindex` allocates an array of zeros with
+a single non-zero element. 
+```julia
+function inplace(array)
```

Comment:
i find it very confusing that this is named `inplace` when it doesn't do anything "inplace"

i think it's to match the heading of this section, which is about "Inplace accumulation", but i think it's more confusing that giving it a name like `sum3` or something that'd you'd actually name a function with this behaviour